### ISSUE TITLE
missDataFun is deparsing entire data frames

### ISSUE THREAD

**Issue by pitakakariki** [lme4/lme4#410]
This can get pretty slow if the data frame is big enough.

The offending line is:

    if (grepl("^\\.\\.[0-9]+$",safeDeparse(ex)))

Would it still work if it were something like one of these?

    if (!is.data.frame(ex) && grepl("^\\.\\.[0-9]+$",safeDeparse(ex)))
    # or
    if (is.symbol(ex) && grepl("^\\.\\.[0-9]+$",safeDeparse(ex)))

Some code that triggers the problem:

```
X <- expand.grid(x1=1:10, x2=1:10, x3=1:10, x4=1:10, g=letters[1:20])

X$y <- X$x1 + rnorm(10)[X$g] + rnorm(200000)
system.time(lmer(y ~ x1 + x2 + x3 + x4 + (1|g), data=X, control=lmerControl(optimizer=NULL)))

f <- function(X) {

    X$y <- X$x1 + rnorm(10)[X$g] + rnorm(200000)

    lmer(y ~ x1 + x2 + x3 + x4 + (1|g), data=X, control=lmerControl(optimizer=NULL))
}

system.time(f(X))

g <- function(X) {

    X$y <- X$x1 + rnorm(10)[X$g] + rnorm(200000)

    lme4:::missDataFun(X)
}

system.time(g(X))
```

**Comment by wahani** [comment]
I believe we (@mgross) have the same issue. Especially with larger data frames this is a problem. Please find below a short reproducible example. We also provide a workaround, which is to reset the environment of the formula passed to lmer:

```r
library(lme4)

lmerTest <- function(f){
  #create simple data set
  n_u <- 1000
  n_y <- 100
  u <- rep(rnorm(n_u), each=n_y)
  id <- rep(1:n_u, each=n_y)
  x <- rnorm(n_u*n_y)
  y <- 1+ 0.5 * x + u + rnorm(n_u*n_y)
  data <- data.frame(y=y, x=x, id=id)
  #slow
  Test1 <- system.time({lmer(f, data = data)})[3]
  #workaround
  Test2 <- system.time({
    environment(f) <- list2env(data)
    lmer(f)
  })[3]
  return(list(times = c(Test1, Test2), data = data))
}

f <- y ~ x+(1|id)
lmerTestExec <- lmerTest(f = f)
lmerTestExec$times
```

We traced the problem back to the same line as already mentioned by @pitakakariki.

**Comment by bbolker** [comment]
Think this is fixed by a1c82fa245d .   Passing tests so far; anyone want to double-check/test on their systems?

**Comment by wahani** [comment]
Yes, the version on the master seems to fix this issue. Thanks!
